client/cpp/synnax/channel/channel.cpp

```diff
@@ -0,0 +1,123 @@
+
+
+Channel::Channel(const api::v1::Channel &ch) {
+    name = ch.name();
+    key = ch.key();
+    index = ch.index();
+    rate = Telem::Rate(ch.rate());
+    is_index = ch.is_index();
+    leaseholder = ch.leaseholder();
+    dataType = Telem::DataType(ch.data_type());
+}
+
+Channel::Channel(
+        std::string name,
+        Telem::DataType dataType,
+        Telem::Rate rate,
+        bool is_index,
+        std::uint32_t leaseholder,
+        ChannelKey index,
+        ChannelKey key
+) : dataType(dataType),
+    name(name),
+    key(key),
+    index(index),
+    rate(rate),
+    is_index(is_index),
+    leaseholder(leaseholder) {}
+
+
+void Channel::to_proto(api::v1::Channel *a) const {
+    a->set_name(name);
+    a->set_data_type(dataType.value);
+    a->set_rate(rate.value);
+    a->set_is_index(is_index);
+    a->set_leaseholder(leaseholder);
+    a->set_index(index);
+    a->set_key(key);
+}
+
+
+Channel ChannelClient::create(
+        std::string name,
+        Telem::DataType data_type,
+        Telem::Rate rate,
+        ChannelKey index,
+        bool is_index
+) {
+    auto req = api::v1::ChannelCreateRequest();
+    auto a = req.add_channels();
+    a->set_name(name);
+    a->set_index(index);
+    a->set_is_index(is_index);
+    a->set_data_type(data_type.value);
+    a->set_rate(rate.value);
+    auto response = create_client->send(CREATE_ENDPOINT, req);
+    return {response.first.channels(0)};
+}
+
+void ChannelClient::create(std::vector<Channel> &channels) {
+    auto req = api::v1::ChannelCreateRequest();
+    for (const auto &ch: channels) {
+        auto a = req.add_channels();
+        ch.to_proto(a);
+    }
+    auto response = create_client->send(CREATE_ENDPOINT, req);
+    for (auto i = 0; i < response.first.channels_size(); i++)
+        channels[i] = Channel(response.first.channels(i));
+}
+
+
+Channel ChannelClient::retrieve(ChannelKey key) {
+    auto req = api::v1::ChannelRetrieveRequest();
+    req.add_keys(key);
+    std::pair<api::v1::ChannelRetrieveResponse, grpc::Status> response =
+            retrieve_client->send(RETRIEVE_ENDPOINT, req);
+    if (response.first.channels_size() == 0)
+        throw QueryError("No channel found with key " + key);
+    return Channel(response.first.channels(0));
+}
+
+Channel ChannelClient::retrieve(const std::string &name) {
+    auto payload = api::v1::ChannelRetrieveRequest();
+    payload.add_names(name);
+    retrieve_client->send(RETRIEVE_ENDPOINT, payload);
+}
+
+std::vector<Channel> ChannelClient::retrieve(const std::vector<ChannelKey> &keys) {
+    auto req = api::v1::ChannelRetrieveRequest();
+    for (auto key: keys)
+        req.add_keys(key);
+    std::pair<api::v1::ChannelRetrieveResponse, grpc::Status> response =
+            retrieve_client->send(RETRIEVE_ENDPOINT, req);
+    std::vector<Channel> channels;
+    for (auto i = 0; i < response.first.channels_size(); i++)
+        channels.emplace_back(response.first.channels(i));
```

Comment:
was this also an IDE suggestion? I also had people telling me `emplace_back` was better than `push_back`, still don't know why...

Reply:
The code actually didn't compile when using `push_back`, still don't completely understand why. I figured out a much bettter way to do this translation so, so resolved.